pallets/did/src/lib.rs

```diff
@@ -316,6 +325,118 @@ where
 	}
 }
 
+/// A web URL starting with either http:// or https://.
+#[derive(Clone, Decode, Debug, Encode, PartialEq)]
+pub struct HttpUrl {
+	payload: Vec<u8>,
+}
+
+impl TryFrom<&[u8]> for HttpUrl {
+	type Error = UrlError;
+
+	// It fails if the byte sequence does not result in an ASCII-encoded string or
+	// if the resulting string contains characters that are not allowed in a URL.
+	fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+		let str_url = str::from_utf8(value).map_err(|_| UrlError::InvalidUrlEncoding)?;
+
+		ensure!(
+			str_url.starts_with("http://") || str_url.starts_with("https://"),
+			UrlError::InvalidUrlScheme
+		);
+
+		ensure!(utils::is_valid_ascii_url(&str_url), UrlError::InvalidUrlEncoding);
+
+		Ok(HttpUrl {
+			payload: value.to_vec(),
+		})
+	}
+}
+
+/// An FTP URL starting with ftp:// or ftps://.
+#[derive(Clone, Decode, Debug, Encode, PartialEq)]
+pub struct FtpUrl {
+	payload: Vec<u8>,
+}
+
+impl TryFrom<&[u8]> for FtpUrl {
+	type Error = UrlError;
+
+	// It fails if the byte sequence does not result in an ASCII-encoded string or
+	// if the resulting string contains characters that are not allowed in a URL.
+	fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+		let str_url = str::from_utf8(value).map_err(|_| UrlError::InvalidUrlEncoding)?;
+
+		ensure!(
+			str_url.starts_with("ftp://") || str_url.starts_with("ftps://"),
+			UrlError::InvalidUrlScheme
+		);
+
+		ensure!(utils::is_valid_ascii_url(&str_url), UrlError::InvalidUrlEncoding);
+
+		Ok(FtpUrl {
+			payload: value.to_vec(),
+		})
+	}
+}
+
+/// An IPFS URL starting with ipfs://. Both CIDs v0 and v1 supported.
+#[derive(Clone, Decode, Debug, Encode, PartialEq)]
+pub struct IpfsUrl {
+	payload: Vec<u8>,
+}
+
+impl TryFrom<&[u8]> for IpfsUrl {
+	type Error = UrlError;
+
+	// It fails if the URL is not ASCII-encoded or does not start with the expected
+	// URL scheme.
+	fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+		let str_url = str::from_utf8(value).map_err(|_| UrlError::InvalidUrlEncoding)?;
+
+		ensure!(str_url.starts_with("ipfs://"), UrlError::InvalidUrlScheme);
+
+		// Remove the first 7 characters (the URL scheme)
+		let slice_to_verify = &str_url[7..];
```

Comment:
This panics if `str_url` is to short. I would rather use [get](https://doc.rust-lang.org/std/primitive.slice.html#method.get) in combination with `.unwrap().expect("The length was ensured by calling starts_with")` or something similar.

Reply:
And maybe even ... have `"ipfs://"` as a const and use the length of that const instead of 7? I think this would make it clear where the 7 comes from even for me! 😅